File: android/app/src/main/java/com/omniagent/assistant/service/FloatingOverlayService.java
package com.omniagent.assistant.service;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.os.IBinder;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;

public class FloatingOverlayService extends Service implements View.OnClickListener, View.OnTouchListener {

    public static final String ACTION_UPDATE_TEXT = "com.omniagent.UPDATE_OVERLAY";
    public static final String EXTRA_TEXT = "extra_text";

    private WindowManager windowManager;
    private View overlayView;
    private TextView statusTextView;
    private TextView arcReactorIcon;
    private WindowManager.LayoutParams windowParams;

    private int initialX;
    private int initialY;
    private float initialTouchX;
    private float initialTouchY;

    public static void updateStatus(Context context, String text) {
        Intent intent = new Intent(context, FloatingOverlayService.class);
        intent.setAction(ACTION_UPDATE_TEXT);
        intent.putExtra(EXTRA_TEXT, text);
        context.startService(intent);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        buildOverlay();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && ACTION_UPDATE_TEXT.equals(intent.getAction())) {
            String text = intent.getStringExtra(EXTRA_TEXT);
            if (statusTextView != null && text != null) {
                statusTextView.setText(text);
            }
        }
        return START_STICKY;
    }

    private void buildOverlay() {
        int layoutType = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O ?
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY :
                WindowManager.LayoutParams.TYPE_PHONE;

        windowParams = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                layoutType,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                PixelFormat.TRANSLUCENT
        );
        windowParams.gravity = Gravity.TOP | Gravity.START;
        windowParams.x = 40;
        windowParams.y = 140;

        // Jarvis Arc Reactor Glowing Pill Container
        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.HORIZONTAL);
        layout.setPadding(28, 16, 28, 16);
        layout.setGravity(Gravity.CENTER_VERTICAL);

        GradientDrawable bg = new GradientDrawable();
        bg.setColor(Color.parseColor("#E6090D16")); // High-tech deep obsidian
        bg.setCornerRadius(45f);
        bg.setStroke(3, Color.parseColor("#00E5FF")); // Jarvis Cyan glow
        layout.setBackground(bg);

        // Arc Reactor Glowing Core Icon
        arcReactorIcon = new TextView(this);
        arcReactorIcon.setId(1001);
        arcReactorIcon.setText("◉ ");
        arcReactorIcon.setTextColor(Color.parseColor("#00E5FF"));
        arcReactorIcon.setTextSize(18f);
        arcReactorIcon.setOnClickListener(this);
        layout.addView(arcReactorIcon);

        statusTextView = new TextView(this);
        statusTextView.setText("JARVIS Online");
        statusTextView.setTextColor(Color.parseColor("#E0F7FA"));
        statusTextView.setTextSize(12f);
        statusTextView.setMaxLines(2);
        layout.addView(statusTextView);

        TextView stopIcon = new TextView(this);
        stopIcon.setId(1002);
        stopIcon.setText("  ✕");
        stopIcon.setTextColor(Color.parseColor("#FF5252"));
        stopIcon.setTextSize(14f);
        stopIcon.setOnClickListener(this);
        layout.addView(stopIcon);

        layout.setOnTouchListener(this);

        overlayView = layout;
        windowManager.addView(overlayView, windowParams);
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == 1001) {
            if (statusTextView != null) statusTextView.setText("Listening, sir...");
            Intent voiceIntent = new Intent(this, VoiceAssistantService.class);
            voiceIntent.setAction(VoiceAssistantService.ACTION_START_LISTEN);
            startService(voiceIntent);
        } else if (id == 1002) {
            if (statusTextView != null) statusTextView.setText("Standby, sir.");
            Intent stopIntent = new Intent(this, VoiceAssistantService.class);
            stopIntent.setAction(VoiceAssistantService.ACTION_STOP);
            startService(stopIntent);
        }
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                initialX = windowParams.x;
                initialY = windowParams.y;
                initialTouchX = event.getRawX();
                initialTouchY = event.getRawY();
                return true;
            case MotionEvent.ACTION_MOVE:
                windowParams.x = initialX + (int) (event.getRawX() - initialTouchX);
                windowParams.y = initialY + (int) (event.getRawY() - initialTouchY);
                windowManager.updateViewLayout(overlayView, windowParams);
                return true;
        }
        return false;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (overlayView != null) {
            windowManager.removeView(overlayView);
            overlayView = null;
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: android/app/src/main/java/com/omniagent/assistant/service/FloatingOverlayService.kt
package com.omniagent.assistant.service

import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PixelFormat
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.os.IBinder
import android.view.*
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

/**
 * Floating Overlay Service (Dynamic Island / Action HUD).
 * Floats on top of all Android applications to provide:
 * 1. 1-tap Voice Command trigger
 * 2. Real-time visual feedback of what the agent is seeing & doing
 * 3. Emergency Cancel / Kill-switch button
 */
class FloatingOverlayService : Service() {

    companion object {
        private const val ACTION_UPDATE_STATUS = "com.omniagent.UPDATE_STATUS"
        private const val EXTRA_STATUS = "extra_status"

        @Volatile
        private var instance: FloatingOverlayService? = null

        fun updateStatus(context: Context, status: String) {
            val intent = Intent(context, FloatingOverlayService::class.java).apply {
                action = ACTION_UPDATE_STATUS
                putExtra(EXTRA_STATUS, status)
            }
            context.startService(intent)
        }

        fun showConfirmationDialog(context: Context, question: String, callback: (Boolean) -> Unit) {
            instance?.displayConfirmation(question, callback)
        }
    }

    private var windowManager: WindowManager? = null
    private var overlayView: View? = null
    private var statusTextView: TextView? = null
    private var micButton: View? = null
    private var stopButton: View? = null

    private var initialX: Int = 0
    private var initialY: Int = 0
    private var initialTouchX: Float = 0f
    private var initialTouchY: Float = 0f

    override fun onCreate() {
        super.onCreate()
        instance = this
        windowManager = getSystemService(WINDOW_SERVICE) as WindowManager
        createFloatingView()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent?.action == ACTION_UPDATE_STATUS) {
            val status = intent.getStringExtra(EXTRA_STATUS) ?: "Idle"
            statusTextView?.text = status
        }
        return START_NOT_STICKY
    }

    private fun createFloatingView() {
        val layoutType = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
        } else {
            @Suppress("DEPRECATION")
            WindowManager.LayoutParams.TYPE_PHONE
        }

        val params = WindowManager.LayoutParams(
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            layoutType,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or
                    WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            PixelFormat.TRANSLUCENT
        ).apply {
            gravity = Gravity.TOP or Gravity.START
            x = 40
            y = 120
        }

        // Programmatic container layout for clean dependency-free overlay
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.HORIZONTAL
            setPadding(24, 16, 24, 16)
            background = GradientDrawable().apply {
                setColor(0xEE0F172A.toInt()) // Deep slate semi-transparent
                cornerRadius = 48f
                setStroke(2, 0x446366F1.toInt()) // Indigo accent border
            }
            gravity = Gravity.CENTER_VERTICAL
            elevation = 16f
        }

        // Mic Button
        val micIcon = TextView(this).apply {
            text = "🎤"
            textSize = 18f
            setPadding(8, 0, 16, 0)
            setOnClickListener {
                val voiceIntent = Intent(this@FloatingOverlayService, VoiceAssistantService::class.java).apply {
                    action = VoiceAssistantService.ACTION_START_LISTENING
                }
                startService(voiceIntent)
            }
        }
        micButton = micIcon
        root.addView(micIcon)

        // Status Text (dynamic status)
        statusTextView = TextView(this).apply {
            text = "OmniAgent Ready"
            setTextColor(Color.WHITE)
            textSize = 13f
            maxLines = 2
            setPadding(0, 0, 16, 0)
        }
        root.addView(statusTextView)

        // Emergency Stop Button
        val stopIcon = TextView(this).apply {
            text = "🛑"
            textSize = 16f
            setPadding(8, 4, 8, 4)
            setOnClickListener {
                statusTextView?.text = "Aborting..."
                val stopIntent = Intent(this@FloatingOverlayService, VoiceAssistantService::class.java).apply {
                    action = VoiceAssistantService.ACTION_STOP_AGENT
                }
                startService(stopIntent)
            }
        }
        stopButton = stopIcon
        root.addView(stopIcon)

        // Touch Listener for dragging the overlay
        root.setOnTouchListener { view, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    initialX = params.x
                    initialY = params.y
                    initialTouchX = event.rawX
                    initialTouchY = event.rawY
                    true
                }
                MotionEvent.ACTION_MOVE -> {
                    params.x = initialX + (event.rawX - initialTouchX).toInt()
                    params.y = initialY + (event.rawY - initialTouchY).toInt()
                    windowManager?.updateViewLayout(view, params)
                    true
                }
                else -> false
            }
        }

        overlayView = root
        windowManager?.addView(root, params)
    }

    fun displayConfirmation(question: String, callback: (Boolean) -> Unit) {
        // Can open an interactive dialog over the screen
        statusTextView?.post {
            statusTextView?.text = "⚠️ $question"
        }
        // Auto-approve after 5s in demo or provide buttons
        callback(true)
    }

    override fun onDestroy() {
        super.onDestroy()
        if (overlayView != null) {
            windowManager?.removeView(overlayView)
            overlayView = null
        }
        instance = null
    }

    override fun onBind(intent: Intent?): IBinder? = null
}
